pwnlib/rop/rop.py

```diff
@@ -1277,6 +1279,116 @@ def search(self, move = 0, regs = None, order = 'size'):
+                for insn in md.disasm(elf.section('.text'), elf.offset_to_vaddr(elf.get_section_by_name('.text').header['sh_offset'])):
+                    if insn.mnemonic == 'lea' and insn.operands[0].reg == X86_REG_R8:
+                        elf.sym['__libc_csu_fini'] = insn.address + insn.size + insn.disp
+                    if insn.mnemonic == 'lea' and insn.operands[0].reg == X86_REG_RCX:
+                        elf.sym['__libc_csu_init'] = insn.address + insn.size + insn.disp
+                        break
+            else:
+                for insn in md.disasm(elf.section('.text'), elf.get_section_by_name('.text').header['sh_addr']):
+                    if insn.mnemonic == 'mov' and insn.operands[0].reg == X86_REG_R8:
+                        elf.sym['__libc_csu_fini'] = insn.operands[1].imm
+                    if insn.mnemonic == 'mov' and insn.operands[0].reg == X86_REG_RCX:
+                        elf.sym['__libc_csu_init'] = insn.operands[1].imm
+                        break
+
+        # Resolve location of _fini address if required
+        if not elf.pie and not call:
+            fini = elf.search(p64(elf.dynamic_by_tag('DT_FINI')['d_ptr'])).next()
+        elif elf.pie and not call:
+            log.error('No non-PIE binaries in [elfs], \'call\' parameter is required')
+
+        csu_function = elf.read(elf.sym['__libc_csu_init'], elf.sym['__libc_csu_fini'] - elf.sym['__libc_csu_init'])
+
+        # 1st gadget: Populate registers in preparation for 2nd gadget
+        for insn in md.disasm(csu_function, elf.sym['__libc_csu_init']):
+            if insn.mnemonic == 'pop' and insn.operands[0].reg == X86_REG_RBX:
+                self.raw(insn.address)
+                break
+        # rbx and rbp must be equal after 'add rbx, 1'
+        self.raw(0x00) # pop rbx
+        self.raw(0x01) # pop rbp
+        if call:
+            self.raw(call) # pop r12
+        else:
+            self.raw(fini) # pop r12
+
+        # Older versions of gcc use r13 to populate rdx then r15d to populate edi, newer versions use the reverse
+        # Account for this when the binary was linked against a glibc that was built with a newer gcc
+        for insn in md.disasm(csu_function, elf.sym['__libc_csu_init']):
+            if insn.mnemonic == 'mov' and insn.operands[0].reg == X86_REG_RDX and insn.operands[1].reg == X86_REG_R13:
+                self.raw(rdx)  # pop r13
+                self.raw(rsi)  # pop r14
+                self.raw(edi)  # pop r15
+                self.raw(insn.address)
+                break
+            elif insn.mnemonic == 'mov' and insn.operands[0].reg == X86_REG_RDX and insn.operands[1].reg == X86_REG_R15:
+                self.raw(edi)  # pop r13
+                self.raw(rsi)  # pop r14
+                self.raw(rdx)  # pop r15
+                self.raw(insn.address)
+                break
+
+        # 2nd gadget: Populate edi, rsi & rdx. Populate optional registers
+        self.raw(0x00) # add rsp, 8
```

Comment:
This should be replaced with padding, like used throughout ROP library, and should use described gadgets